# Unit 2 Assessment - Monte Carlo Simulations

## Instructions

**This is a practice assessment; however, the instructions below are similar to what you will see on the actual assessment. There are some noteable differences. (1) The practice assessment is sometimes longer than an actual assessment. (2) The actual assessment will have both an individual part and a group part, which you will solve with an assigned partner.**

In this assessment, you are going to investigate models that use random numbers. 

Scores are determined by:

- Successfully starting the C Level = 50 pts
- Perfectly completing the C Level = 75 pts
- Perfectly completing the B and C Levels = 85 pts
- Perfectly completing the A, B, and C Levels = 100 pts

You may use your Colab notebooks, our textbook, my notebook solutions, and any links to web sites I provide. (You may not use any other person or web site or book or resource, in general.) 

You may ask me for help **once**; however, you may ask for clarification as often as needed.

Add additional cells for both code and markdown as needed. Write answers to questions in narrative form in markdown. You may print values you need in your code, and then use these values in a written response.

All graphs should have correct titles and axis labels (with units).

## Grade

<font color="green"></font>

Level | Grade | Comment
--- | --- | ---
C (75 pts) | | 
B (10 pts) | | 
A (15 pts) | | 
Total | 


# Solutions

## Level C

### Exercise 0

1. Save a copy of this notebook to Google Drive. Then click "Share" and share it with "hpuphysics@gmail.com". Make sure I have permission to edit your document.

2. Add a text cell above and type your name as a level one heading in markdown. (A level one heading starts with # on its own line.)

3. Run the `import` statements below to add packages.

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import random as rand

### Exercise 1

Print a random number in the interval $[0,1)$.

In [3]:
r = rand.random()
print(r)

0.8768187669660349


### Exercise 2

Print a random integer in the interval $[1,6]$.

In [4]:
r = rand.randint(1,6)
print(r)

2


### Exercise 3

Suppose you have a four-sided di with the numbers 1, 2, 3, and 4. How would a four-sided die (with equal probability of each side) be shaped? You can just describe it. You don't have to know the formal name.

It would be shaped like a triangular pyramid with four equilaterial triangles, three on the sides and one on the bottom. To know what you rolled, you would have to pick it up and look at the bottom. Weird, I know.

### Exercise 4

Write a program to roll a four-sided di 100 times and count the number of times you roll a 3. (Print the final result.)

In [6]:
N = 100

count = 0

for i in range(N):
    r = rand.randint(1,4)
    if r == 3:
        count = count + 1

print("count = ", count)

count =  30


### Exercise 5

Copy and paste (below) your program from Exercise 4. Adjust your program as necessary to determine the probability of rolling a 3. Compare the result to what you expect from probability theory.

In [11]:
N = 1000000

count = 0

for i in range(N):
    r = rand.randint(1,4)
    if r == 3:
        count = count + 1

p = count/N*100
print("percentage of rolls = ", p)

percentage of rolls =  25.0519


To determine the probability of rolling a 3, I calculated and printed the percentage of rolls that are a 3 for a very large number of rolls. The probability, to two significant figures, is 25%. For a four-sided di, the probability of rolling a 3 is theoretically 1/4 = 25%.

### Exercise 6

Suppose you roll two four-sided dice and record the sum (total) of the two dice. How many unique combinations of the dice are there and how many unique outcomes (for the sum) exist for the total of two four-sided dice?

Here is a table of results

di A | di B | total
--- | --- | ---
1 | 1 | 2
1 | 2 | 3
1 | 3 | 4
1 | 4 | 5
2 | 1 | 3
2 | 2 | 4
2 | 3 | 5
2 | 4 | 6
3 | 1 | 4
3 | 2 | 5
3 | 3 | 6
3 | 4 | 7
4 | 1 | 5
4 | 2 | 6
4 | 3 | 7
4 | 4 | 8

There are 16 possible rolls of the dice. Only one outcome gives a total of 2, and one outcome gives a total of 8. The number of ways each outcome can be rolled are shown below.

total for two dice | number of ways | probability
--- | --- | ---
2 | 1 | 1/16 = 6.25%
3 | 2 | 2/16 = 12.5%
4 | 3 | 3/16 = 18.75%
5 | 4 | 4/16 = 25%
6 | 3 | 3/16 = 18.75%
7 | 2 | 2/16 = 12.5%
8 | 1 | 1/16 = 6.25%

Note that if you add the number of ways to get each total (or outcome), you get 16, as expected.

### Exercise 7

Write a program to roll two four-sided dice. Calculate and print the percentage of rolls that produces each outcome (i.e. total). Compare your results to your table in the previous exercise. (Note: it is acceptable to write a single program that works for one outcome, record the result, and then edit your program and run it again.)

In [28]:
N = 100000

count = 0

for i in range(N):
    r1 = rand.randint(1,4)
    r2 = rand.randint(1,4)
    sum = r1 + r2
    if sum == 8:
        count = count + 1

p = count/N*100
print("probability = ", p)

probability =  6.337


I changed line 9 to test for each possible sum, from 2 to 8. I used 100000 rolls.

The results are:

total for two dice | probability
--- | ---
2 | 6.28%
3 | 12.5%
4 | 18.6%
5 | 25.0%
6 | 18.7%
7 | 12.4%
8 | 6.3%

## Level B

### Exercise 1

Print a random number from a gaussian distribution with a mean of 100 and standard deviation 10.

In [31]:
r = rand.gauss(100, 10)
print(r)

95.40317319633638


### Exercise 2

There is a philosophy in business called [six sigma](https://en.wikipedia.org/wiki/Six_Sigma). According to Wikipedia, six sigma "is a set of techniques and tools for process improvement. It was introduced by American engineer Bill Smith while working at Motorola in 1986. Jack Welch made it central to his business strategy at General Electric in 1995. A six sigma process is one in which 99.99966% of all opportunities to produce some feature of a part are statistically expected to be free of defects."

Re-read that last sentence, "A six sigma process is one in which 99.99966% of all opportunities to produce some feature of a part are statistically expected to be free of defects." That means that a defect will only occur in 34 out of 10 million parts! Where does this number come from? Let's simulate it. **At first, we will only use one sigma. Then we will increase it later.** 

(a) Write a loop that selects a random number from a gaussian distribution with a mean of 100 and standard deviation of 10. If a random number is greater than the mean minus sigma OR if the random number is less than the mean plus sigma, count the occurrence of this number.

Note that if your variables for mean and standard deviation are named `mean` and `stdev` respectively, then an `if` statement that checks whether a random number `r` is greater than the mean minus sigma AND less than the mean plus sigma would be:

```python
if r > mean - sigma and r < mean + sigma:
```

Based on your program, what is the probability that that a random number from a gaussian distribution falls within the range of mean $\pm$ sigma?


In [42]:
N = 1000000

mean = 100
sigma = 10
count = 0

for i in range(N):
    r = rand.gauss(mean, sigma)
    if r > mean - sigma and r < mean + sigma:
        count = count + 1
        
p = count/N*100
print("probability = ", p)

probability =  68.2218


For one million trials, the probability is 68.2%

(b) Theoretically, we expect 68.3% of the data to fall within one standard deviation of the mean. How does your result from the Monte Carlo simulation compare to theory?

For one million trials, running my program multiple times gives results between 68.2% and 68.3%. The numerical approximation agrees well with theory.

(c) Copy your program and paste it below. Using your model, compute the probability a random number from a Gaussian distribution falls within two sigma of the mean.

In [46]:
N = 1000000

mean = 100
sigma = 10
count = 0

for i in range(N):
    r = rand.gauss(mean, sigma)
    if r > mean - 2*sigma and r < mean + 2*sigma:
        count = count + 1
        
p = count/N*100
print("probability = ", p)

probability =  95.4324


(d) With calculus, it is possible to derive that we expect 95.5% percent of the data to fall within two standard deviations. How does the result you found compare to the theoretical result?

For one million trials, running my program multiple times gives results between 95.4% and 95.5%. The numerical approximation agrees well with theory.

## Level A

### Exercise 1

A tossed fair coin has a 50% chance (probability = 0.5) of being heads. Write a program to simulate tossing a fair coin and calculate (and print) the percentage of $N$ tosses that are heads.

In [47]:
N = 1000000

count = 0

for i in range(N):
    r = rand.random()
    if r < 0.5: #heads
        count = count + 1

p = count/N*100
print("percentage = ", p)

percentage =  50.0225


### Exercise 2

Suppose we have an unfair coin that is weighted in some way such that the probability of being heads is 0.3. Copy your program from the previous exercise and paste it below. Edit it so that the probability of being heads is 0.3, and calculate (and print) the percentage of $N$ tosses that are heads.

In [48]:
N = 1000000

count = 0

for i in range(N):
    r = rand.random()
    if r < 0.3: #heads
        count = count + 1

p = count/N*100
print("percentage = ", p)

percentage =  30.024099999999997


### Exercise 3

Beth Harmon, the fictious character in the mini-series *Queen's Gambit* has agreed to a particular challenge. She will play two Russian players whom I will call A and C in three alternating games of chess that can be player A, then C, then A again (sequence $ACA$) or player C, then A, then C again (sequence $CAC$). Not only must she win 2 out of 3 games, she must also with two *in a row*.

Player C is better than player A, and Beth figures that her probability of beating player A is 0.9, and her probability of beating player C is 0.8.

The question is: **Should she choose to play them in the sequence $ACA$ or the sequence $CAC$?**

I've written the function below to take a sequence (`ACA` or `CAC` as a string), and determine whether Beth wins two chess games in a row. The function returns a 1 if Beth wins two in a row, and the function returns a 0 if Beth does not win two in a row. The variables `p` and `q` define the probability of beating player A and player C, respectively.

In [58]:
def sequence(s):
    #s should be "ACA" or "CAC"
    p = 0.9 #probability of beating A
    q = 0.8 #probability of beating C
    
    result = 0 #the result is whether she wins two in a row or not
        
    if s == "ACA":
        games = [] #store results of a single game as 0 or 1
        # play player A
        r = rand.random()
        if r < p: #Beth wins
            games.append(1)
        else: #Beth loses
            games.append(0)

        # play player C
        r = rand.random()
        if r < q: #Beth wins
            games.append(1)
        else: #Beth loses
            games.append(0)

        # play player A
        r = rand.random()
        if r < p: #Beth wins
            games.append(1)
        else: #Beth loses
            games.append(0)

    if s == "CAC":
        games = [] #store results of a single game as 0 or 1
        # play player C
        r = rand.random()
        if r < q: #Beth wins
            games.append(1)
        else: #Beth loses
            games.append(0)

        # play player A
        r = rand.random()
        if r < p: #Beth wins
            games.append(1)
        else: #Beth loses
            games.append(0)

        # play player C
        r = rand.random()
        if r < q: #Beth wins
            games.append(1)
        else: #Beth loses
            games.append(0)
            
    
    #determine if Beth wins two games in a row
    if games[0] == 1 and games[1] == 1:
        result = 1
    elif games[1] == 1 and games[2] == 1:
        result = 1
    else:
        result = 0
    
    return result

The progam below plays one sequence of three chess games and tells you (with a 1 or 0) whether Beth wins (1) or not (0). Run it to see how it works. Note that you can rerun this cell over and over to see if the result changes. Also, you can try the sequence `"ACA"` or the sequence `CAC`.

In [59]:
outcome = sequence("ACA")
print("outcome = ", outcome)

outcome =  1


Write a loop to answer the following question. What is the probability that Beth wins if she plays the sequence $ACA$?

In [60]:
N = 100000

count = 1

for i in range(N):
    outcome = sequence("ACA")
    
    if outcome == 1:
        count = count + 1

p = count/N*100

print("probability of winning is ", p)

probability of winning is  79.041


Write a loop to answer the following question. What is the probability that Beth wins if she plays the sequence $CAC$?

In [61]:
N = 100000

count = 1

for i in range(N):
    outcome = sequence("CAC")
    
    if outcome == 1:
        count = count + 1

p = count/N*100

print("probability of winning is ", p)

probability of winning is  86.36200000000001


If Beth wants to maximize her probability of winning, which sequence should she choose?

Beth is more like to win if she plays the more difficult player first and last. So she has to play the more difficult player twice. But this actually increases her probability of winning two in a row. It's very counterintuitive!